{
  "cells": [
    {
      "cell_type": "markdown",
      "metadata": {},
      "source": [
        "# Usecase 3: Microbiome load prediction data preparation"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {},
      "source": [
        "This notebook prepares the dataset for the microbiome load prediction usecase following the general data preparation approach outlined in [the original publication by Nishijima et al. 2024](10.1016/j.cell.2024.10.022). It can be run in the following conda environment:\n",
        "\n",
        "This notebook can be run in the following conda environment (last command must be launched from root of his repos):\n",
        "```shell\n",
        "mamba env create -f environment_prep_data.yml\n",
        "conda activate ritme_examples_prep_data\n",
        "pip install -e .\n",
        "qiime dev refresh-cache\n",
        "```"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {},
      "source": [
        "## Setup"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 1,
      "metadata": {},
      "outputs": [
        {
          "name": "stderr",
          "output_type": "stream",
          "text": [
            "/cluster/project/bokulich/adamova/_conda_envs/ritme_examples_prep_data/lib/python3.9/site-packages/qiime2/core/archive/provenance.py:13: UserWarning: pkg_resources is deprecated as an API. See https://setuptools.pypa.io/en/latest/pkg_resources.html. The pkg_resources package is slated for removal as early as 2025-11-30. Refrain from using this package or pin to Setuptools<81.\n",
            "  import pkg_resources\n"
          ]
        }
      ],
      "source": [
        "import numpy as np\n",
        "import pandas as pd\n",
        "import qiime2 as q2\n",
        "\n",
        "from src.process_u3 import process_feature_table\n",
        "\n",
        "%load_ext autoreload\n",
        "%autoreload 2\n",
        "\n",
        "%matplotlib inline"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {},
      "source": [
        "## Fetch data"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 2,
      "metadata": {
        "vscode": {
          "languageId": "shellscript"
        }
      },
      "outputs": [
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "Data already fetched in ../../data/u3_mlp_nishijima24, skipping.\n"
          ]
        }
      ],
      "source": [
        "! ./../../src/fetch_mlp_data.sh"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 3,
      "metadata": {},
      "outputs": [],
      "source": [
        "path_to_data = \"../../data/u3_mlp_nishijima24\""
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {},
      "source": [
        "## Create Galaxy dataset"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {},
      "source": [
        "### Metadata"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 4,
      "metadata": {},
      "outputs": [
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "(1894, 3)\n"
          ]
        },
        {
          "data": {
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>count</th>\n",
              "      <th>cohort</th>\n",
              "      <th>count_log10</th>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>ID</th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>ALCO_PYGJU</th>\n",
              "      <td>5.904665e+10</td>\n",
              "      <td>AlcoChallenge</td>\n",
              "      <td>10.771195</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>ALCO_ZYTNB</th>\n",
              "      <td>8.014375e+10</td>\n",
              "      <td>AlcoChallenge</td>\n",
              "      <td>10.903870</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>ALCO_QXCHU</th>\n",
              "      <td>5.972596e+10</td>\n",
              "      <td>AlcoChallenge</td>\n",
              "      <td>10.776163</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>ALCO_GNRDS</th>\n",
              "      <td>8.725575e+10</td>\n",
              "      <td>AlcoChallenge</td>\n",
              "      <td>10.940794</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>ALCO_YVDGT</th>\n",
              "      <td>5.420986e+10</td>\n",
              "      <td>AlcoChallenge</td>\n",
              "      <td>10.734078</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "</div>"
            ],
            "text/plain": [
              "                   count         cohort  count_log10\n",
              "ID                                                  \n",
              "ALCO_PYGJU  5.904665e+10  AlcoChallenge    10.771195\n",
              "ALCO_ZYTNB  8.014375e+10  AlcoChallenge    10.903870\n",
              "ALCO_QXCHU  5.972596e+10  AlcoChallenge    10.776163\n",
              "ALCO_GNRDS  8.725575e+10  AlcoChallenge    10.940794\n",
              "ALCO_YVDGT  5.420986e+10  AlcoChallenge    10.734078"
            ]
          },
          "execution_count": 4,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "galaxy_md = pd.read_csv(f\"{path_to_data}/GALAXY_load.tsv\", sep=\"\\t\", index_col=0)\n",
        "galaxy_md[\"count_log10\"] = np.log10(galaxy_md[\"count\"])\n",
        "\n",
        "print(galaxy_md.shape)\n",
        "galaxy_md.head()"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 5,
      "metadata": {},
      "outputs": [],
      "source": [
        "# save to disk\n",
        "galaxy_md.to_csv(f\"{path_to_data}/md_galaxy.tsv\", sep=\"\\t\")"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {},
      "source": [
        "### Feature table"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 6,
      "metadata": {},
      "outputs": [
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "Original shape (1894, 14213)\n",
            "(1894, 14213)\n"
          ]
        }
      ],
      "source": [
        "galaxy_motus = process_feature_table(path_to_data, \"GALAXY_mOTUs_v25\")\n",
        "print(galaxy_motus.shape)\n",
        "\n",
        "# # save to disk\n",
        "# galaxy_motus.to_csv(f\"{path_to_data}/galaxy_otu_table.tsv\", sep=\"\\t\")\n",
        "# galaxy_motus.head()"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 7,
      "metadata": {},
      "outputs": [
        {
          "data": {
            "text/plain": [
              "{12, 18, 19}"
            ]
          },
          "execution_count": 7,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "# ls_tup = [(x, len(x)) for x in galaxy_motus.columns.tolist()]\n",
        "set([len(x) for x in galaxy_motus.columns.tolist()])"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 8,
      "metadata": {},
      "outputs": [],
      "source": [
        "# are they relative abundances?\n",
        "assert galaxy_motus.sum(axis=1).round(4).eq(1.0).all()"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 9,
      "metadata": {},
      "outputs": [],
      "source": [
        "# check are all sample IDs present in metadata?\n",
        "assert len([x for x in galaxy_motus.index if x not in galaxy_md.index]) == 0\n",
        "assert len([x for x in galaxy_md.index if x not in galaxy_motus.index]) == 0"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {},
      "source": [
        "### Taxonomy"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 10,
      "metadata": {},
      "outputs": [
        {
          "data": {
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>Kingdom</th>\n",
              "      <th>Phylum</th>\n",
              "      <th>Class</th>\n",
              "      <th>Order</th>\n",
              "      <th>Family</th>\n",
              "      <th>Genus</th>\n",
              "      <th>Species</th>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>mOTU</th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>Bacteria</td>\n",
              "      <td>Spirochaetota</td>\n",
              "      <td>Leptospirae</td>\n",
              "      <td>Leptospirales</td>\n",
              "      <td>Leptospiraceae</td>\n",
              "      <td>Leptospira</td>\n",
              "      <td>Leptospira_alexanderi</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2</th>\n",
              "      <td>Bacteria</td>\n",
              "      <td>Spirochaetota</td>\n",
              "      <td>Leptospirae</td>\n",
              "      <td>Leptospirales</td>\n",
              "      <td>Leptospiraceae</td>\n",
              "      <td>Leptospira</td>\n",
              "      <td>Leptospira_weilii</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3</th>\n",
              "      <td>Bacteria</td>\n",
              "      <td>Spirochaetota</td>\n",
              "      <td>Leptospirae</td>\n",
              "      <td>Leptospirales</td>\n",
              "      <td>Leptospiraceae</td>\n",
              "      <td>Leptospira</td>\n",
              "      <td>Leptospira_weilii</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4</th>\n",
              "      <td>Bacteria</td>\n",
              "      <td>Bacteroidota</td>\n",
              "      <td>Bacteroidia</td>\n",
              "      <td>Flavobacteriales</td>\n",
              "      <td>Weeksellaceae</td>\n",
              "      <td>Chryseobacterium</td>\n",
              "      <td>Chryseobacterium_rhizosphaerae</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>5</th>\n",
              "      <td>Bacteria</td>\n",
              "      <td>Bacteroidota</td>\n",
              "      <td>Bacteroidia</td>\n",
              "      <td>Flavobacteriales</td>\n",
              "      <td>Weeksellaceae</td>\n",
              "      <td>Chryseobacterium</td>\n",
              "      <td>Chryseobacterium_gallinarum</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "</div>"
            ],
            "text/plain": [
              "       Kingdom         Phylum        Class             Order          Family  \\\n",
              "mOTU                                                                           \n",
              "1     Bacteria  Spirochaetota  Leptospirae     Leptospirales  Leptospiraceae   \n",
              "2     Bacteria  Spirochaetota  Leptospirae     Leptospirales  Leptospiraceae   \n",
              "3     Bacteria  Spirochaetota  Leptospirae     Leptospirales  Leptospiraceae   \n",
              "4     Bacteria   Bacteroidota  Bacteroidia  Flavobacteriales   Weeksellaceae   \n",
              "5     Bacteria   Bacteroidota  Bacteroidia  Flavobacteriales   Weeksellaceae   \n",
              "\n",
              "                 Genus                         Species  \n",
              "mOTU                                                    \n",
              "1           Leptospira           Leptospira_alexanderi  \n",
              "2           Leptospira               Leptospira_weilii  \n",
              "3           Leptospira               Leptospira_weilii  \n",
              "4     Chryseobacterium  Chryseobacterium_rhizosphaerae  \n",
              "5     Chryseobacterium     Chryseobacterium_gallinarum  "
            ]
          },
          "execution_count": 10,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "taxonomy_mapping = pd.read_csv(\n",
        "    \"../../data/u3_mlp_nishijima24/motus2GTDB.txt\", sep=\"\\t\", index_col=0\n",
        ")\n",
        "\n",
        "# remove empty spaces from column values\n",
        "for col in taxonomy_mapping.columns:\n",
        "    taxonomy_mapping[col] = taxonomy_mapping[col].str.replace(\" \", \"_\")\n",
        "\n",
        "taxonomy_mapping.head()"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 11,
      "metadata": {},
      "outputs": [
        {
          "data": {
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>Taxon</th>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>Feature ID</th>\n",
              "      <th></th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>ref_mOTU_v25_00001</th>\n",
              "      <td>k__Bacteria; p__Spirochaetota; c__Leptospirae;...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>ref_mOTU_v25_00002</th>\n",
              "      <td>k__Bacteria; p__Spirochaetota; c__Leptospirae;...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>ref_mOTU_v25_00003</th>\n",
              "      <td>k__Bacteria; p__Spirochaetota; c__Leptospirae;...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>ref_mOTU_v25_00004</th>\n",
              "      <td>k__Bacteria; p__Bacteroidota; c__Bacteroidia; ...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>ref_mOTU_v25_00005</th>\n",
              "      <td>k__Bacteria; p__Bacteroidota; c__Bacteroidia; ...</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "</div>"
            ],
            "text/plain": [
              "                                                                Taxon\n",
              "Feature ID                                                           \n",
              "ref_mOTU_v25_00001  k__Bacteria; p__Spirochaetota; c__Leptospirae;...\n",
              "ref_mOTU_v25_00002  k__Bacteria; p__Spirochaetota; c__Leptospirae;...\n",
              "ref_mOTU_v25_00003  k__Bacteria; p__Spirochaetota; c__Leptospirae;...\n",
              "ref_mOTU_v25_00004  k__Bacteria; p__Bacteroidota; c__Bacteroidia; ...\n",
              "ref_mOTU_v25_00005  k__Bacteria; p__Bacteroidota; c__Bacteroidia; ..."
            ]
          },
          "execution_count": 11,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "prefix_matching = {\n",
        "    \"Kingdom\": \"k__\",\n",
        "    \"Phylum\": \"p__\",\n",
        "    \"Class\": \"c__\",\n",
        "    \"Order\": \"o__\",\n",
        "    \"Family\": \"f__\",\n",
        "    \"Genus\": \"g__\",\n",
        "    \"Species\": \"s__\",\n",
        "}\n",
        "\n",
        "tax_df = pd.DataFrame(index=taxonomy_mapping.index)\n",
        "tax_df[\"Taxon\"] = taxonomy_mapping.apply(\n",
        "    lambda x: \"; \".join(\n",
        "        [f\"{prefix_matching[k]}{v}\" for k, v in x.items() if not pd.isna(v)]\n",
        "    ),\n",
        "    axis=1,\n",
        ")\n",
        "# create correct index\n",
        "tax_df.index = [f\"ref_mOTU_v25_{int(x):05d}\" for x in tax_df.index.tolist()]\n",
        "tax_df.index.name = \"Feature ID\"\n",
        "\n",
        "# add unclassified\n",
        "unknown_class = (\n",
        "    \"k__undef; p__undef; c__undef; o__undef; f__undef; g__undef; s__undef\"\n",
        ")\n",
        "tax_df.loc[\"unclassified\", \"Taxon\"] = unknown_class\n",
        "\n",
        "no_tax_given = [x for x in galaxy_motus.columns.tolist() if x not in tax_df.index]\n",
        "for x in no_tax_given:\n",
        "    tax_df.loc[x, \"Taxon\"] = unknown_class\n",
        "tax_df.head()"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 12,
      "metadata": {},
      "outputs": [
        {
          "data": {
            "text/plain": [
              "'../../data/u3_mlp_nishijima24/u3_taxonomy.qza'"
            ]
          },
          "execution_count": 12,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "# save to disk\n",
        "tax_art = q2.Artifact.import_data(\"FeatureData[Taxonomy]\", tax_df)\n",
        "tax_art.save(f\"{path_to_data}/u3_taxonomy.qza\")"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {},
      "source": [
        "No phylogeny tree can be constructed since we do not have the nucleotide sequences of these mOTUs -> no trac trainable"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {},
      "source": [
        "## Create Metacardis dataset"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {},
      "source": [
        "### Metadata"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 13,
      "metadata": {},
      "outputs": [
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "(1812, 3)\n"
          ]
        },
        {
          "data": {
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>count</th>\n",
              "      <th>cohort</th>\n",
              "      <th>count_log10</th>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>ID</th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>M0x10MCx1134</th>\n",
              "      <td>1.961080e+11</td>\n",
              "      <td>Severe/morbid obesity</td>\n",
              "      <td>11.292495</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>M0x10MCx1135</th>\n",
              "      <td>1.432750e+11</td>\n",
              "      <td>Type II diabetes</td>\n",
              "      <td>11.156170</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>M0x10MCx1138</th>\n",
              "      <td>7.262570e+10</td>\n",
              "      <td>Type II diabetes</td>\n",
              "      <td>10.861090</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>M0x10MCx1140</th>\n",
              "      <td>7.115385e+10</td>\n",
              "      <td>Severe/morbid obesity</td>\n",
              "      <td>10.852198</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>M0x10MCx1143</th>\n",
              "      <td>3.603352e+10</td>\n",
              "      <td>Type II diabetes</td>\n",
              "      <td>10.556707</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "</div>"
            ],
            "text/plain": [
              "                     count                 cohort  count_log10\n",
              "ID                                                            \n",
              "M0x10MCx1134  1.961080e+11  Severe/morbid obesity    11.292495\n",
              "M0x10MCx1135  1.432750e+11       Type II diabetes    11.156170\n",
              "M0x10MCx1138  7.262570e+10       Type II diabetes    10.861090\n",
              "M0x10MCx1140  7.115385e+10  Severe/morbid obesity    10.852198\n",
              "M0x10MCx1143  3.603352e+10       Type II diabetes    10.556707"
            ]
          },
          "execution_count": 13,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "metacardis_md = pd.read_csv(\n",
        "    f\"{path_to_data}/MetaCardis_load.tsv\", sep=\"\\t\", index_col=0\n",
        ")\n",
        "\n",
        "# according to publication perform log10 transformation\n",
        "metacardis_md[\"count_log10\"] = np.log10(metacardis_md[\"count\"])\n",
        "\n",
        "print(metacardis_md.shape)\n",
        "metacardis_md.head()"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 14,
      "metadata": {},
      "outputs": [],
      "source": [
        "# save to disk\n",
        "metacardis_md.to_csv(f\"{path_to_data}/md_metacardis.tsv\", sep=\"\\t\")"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {},
      "source": [
        "### Feature table"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 15,
      "metadata": {},
      "outputs": [
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "Original shape (1812, 14213)\n",
            "(1812, 14213)\n"
          ]
        },
        {
          "data": {
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>ref_mOTU_v25_10354</th>\n",
              "      <th>unclassified</th>\n",
              "      <th>ref_mOTU_v25_04788</th>\n",
              "      <th>ref_mOTU_v25_03694</th>\n",
              "      <th>ref_mOTU_v25_06702</th>\n",
              "      <th>ref_mOTU_v25_10941</th>\n",
              "      <th>ref_mOTU_v25_10763</th>\n",
              "      <th>ref_mOTU_v25_06703</th>\n",
              "      <th>ref_mOTU_v25_10828</th>\n",
              "      <th>ref_mOTU_v25_04651</th>\n",
              "      <th>...</th>\n",
              "      <th>ref_mOTU_v25_08216</th>\n",
              "      <th>ref_mOTU_v25_03508</th>\n",
              "      <th>ref_mOTU_v25_08217</th>\n",
              "      <th>ref_mOTU_v25_08215</th>\n",
              "      <th>ref_mOTU_v25_07201</th>\n",
              "      <th>ref_mOTU_v25_10802</th>\n",
              "      <th>ref_mOTU_v25_10650</th>\n",
              "      <th>ref_mOTU_v25_11363</th>\n",
              "      <th>ref_mOTU_v25_07158</th>\n",
              "      <th>ref_mOTU_v25_07159</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>M0x10MCx1134</th>\n",
              "      <td>0</td>\n",
              "      <td>0.035435</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>...</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>M0x10MCx1135</th>\n",
              "      <td>0</td>\n",
              "      <td>0.032264</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>...</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>M0x10MCx1138</th>\n",
              "      <td>0</td>\n",
              "      <td>0.027916</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>...</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>M0x10MCx1140</th>\n",
              "      <td>0</td>\n",
              "      <td>0.042573</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>...</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>M0x10MCx1143</th>\n",
              "      <td>0</td>\n",
              "      <td>0.022910</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>...</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "<p>5 rows Ã— 14213 columns</p>\n",
              "</div>"
            ],
            "text/plain": [
              "              ref_mOTU_v25_10354  unclassified  ref_mOTU_v25_04788  \\\n",
              "M0x10MCx1134                   0      0.035435                 0.0   \n",
              "M0x10MCx1135                   0      0.032264                 0.0   \n",
              "M0x10MCx1138                   0      0.027916                 0.0   \n",
              "M0x10MCx1140                   0      0.042573                 0.0   \n",
              "M0x10MCx1143                   0      0.022910                 0.0   \n",
              "\n",
              "              ref_mOTU_v25_03694  ref_mOTU_v25_06702  ref_mOTU_v25_10941  \\\n",
              "M0x10MCx1134                 0.0                   0                   0   \n",
              "M0x10MCx1135                 0.0                   0                   0   \n",
              "M0x10MCx1138                 0.0                   0                   0   \n",
              "M0x10MCx1140                 0.0                   0                   0   \n",
              "M0x10MCx1143                 0.0                   0                   0   \n",
              "\n",
              "              ref_mOTU_v25_10763  ref_mOTU_v25_06703  ref_mOTU_v25_10828  \\\n",
              "M0x10MCx1134                   0                   0                   0   \n",
              "M0x10MCx1135                   0                   0                   0   \n",
              "M0x10MCx1138                   0                   0                   0   \n",
              "M0x10MCx1140                   0                   0                   0   \n",
              "M0x10MCx1143                   0                   0                   0   \n",
              "\n",
              "              ref_mOTU_v25_04651  ...  ref_mOTU_v25_08216  ref_mOTU_v25_03508  \\\n",
              "M0x10MCx1134                   0  ...                   0                   0   \n",
              "M0x10MCx1135                   0  ...                   0                   0   \n",
              "M0x10MCx1138                   0  ...                   0                   0   \n",
              "M0x10MCx1140                   0  ...                   0                   0   \n",
              "M0x10MCx1143                   0  ...                   0                   0   \n",
              "\n",
              "              ref_mOTU_v25_08217  ref_mOTU_v25_08215  ref_mOTU_v25_07201  \\\n",
              "M0x10MCx1134                   0                   0                   0   \n",
              "M0x10MCx1135                   0                   0                   0   \n",
              "M0x10MCx1138                   0                   0                   0   \n",
              "M0x10MCx1140                   0                   0                   0   \n",
              "M0x10MCx1143                   0                   0                   0   \n",
              "\n",
              "              ref_mOTU_v25_10802  ref_mOTU_v25_10650  ref_mOTU_v25_11363  \\\n",
              "M0x10MCx1134                   0                   0                   0   \n",
              "M0x10MCx1135                   0                   0                   0   \n",
              "M0x10MCx1138                   0                   0                   0   \n",
              "M0x10MCx1140                   0                   0                   0   \n",
              "M0x10MCx1143                   0                   0                   0   \n",
              "\n",
              "              ref_mOTU_v25_07158  ref_mOTU_v25_07159  \n",
              "M0x10MCx1134                   0                   0  \n",
              "M0x10MCx1135                   0                   0  \n",
              "M0x10MCx1138                   0                   0  \n",
              "M0x10MCx1140                   0                   0  \n",
              "M0x10MCx1143                   0                   0  \n",
              "\n",
              "[5 rows x 14213 columns]"
            ]
          },
          "execution_count": 15,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "metacardis_motus = process_feature_table(path_to_data, \"MetaCardis_mOTUs_v25\")\n",
        "print(metacardis_motus.shape)\n",
        "\n",
        "# save to disk\n",
        "metacardis_motus.to_csv(f\"{path_to_data}/metacardis_otu_table.tsv\", sep=\"\\t\")\n",
        "\n",
        "metacardis_motus.head()"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 16,
      "metadata": {},
      "outputs": [],
      "source": [
        "# are they relative abundances?\n",
        "assert metacardis_motus.sum(axis=1).round(4).eq(1.0).all()"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 17,
      "metadata": {},
      "outputs": [],
      "source": [
        "# check are all sample IDs present in metadata?\n",
        "assert len([x for x in metacardis_motus.index if x not in metacardis_md.index]) == 0\n",
        "assert len([x for x in metacardis_md.index if x not in metacardis_motus.index]) == 0"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 18,
      "metadata": {},
      "outputs": [
        {
          "data": {
            "text/plain": [
              "{12, 18, 19}"
            ]
          },
          "execution_count": 18,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "set([len(x) for x in metacardis_motus.columns.tolist()])"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {},
      "source": [
        "### Taxonomy"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {},
      "source": [
        "was already processed above as `tax_art` - same mapping can be used for both datasets."
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {},
      "source": [
        "No phylogeny tree can be constructed since we do not have the nucleotide sequences of these mOTUs -> no trac trainable"
      ]
    }
  ],
  "metadata": {
    "kernelspec": {
      "display_name": "ritme_examples_prep_data",
      "language": "python",
      "name": "python3"
    },
    "language_info": {
      "codemirror_mode": {
        "name": "ipython",
        "version": 3
      },
      "file_extension": ".py",
      "mimetype": "text/x-python",
      "name": "python",
      "nbconvert_exporter": "python",
      "pygments_lexer": "ipython3",
      "version": "3.9.23"
    }
  },
  "nbformat": 4,
  "nbformat_minor": 2
}
