{
  "cells": [
    {
      "cell_type": "markdown",
      "metadata": {},
      "source": [
        "# Usecase 3: Microbiome load prediction data preparation"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {},
      "source": [
        "This notebook prepares the dataset for the microbiome load prediction usecase following the general data preparation approach outlined in [the original publication by Nishijima et al. 2024](10.1016/j.cell.2024.10.022). It can be run in the following conda environment:\n",
        "\n",
        "This notebook can be run in the following conda environment:\n",
        "```shell\n",
        "mamba env create -f environment_prep_data.yml\n",
        "conda activate ritme_examples_prep_data\n",
        "pip install -e .\n",
        "qiime dev refresh-cache\n",
        "```"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {},
      "source": [
        "## Setup"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {},
      "outputs": [],
      "source": [
        "import os\n",
        "import re\n",
        "import subprocess\n",
        "\n",
        "import numpy as np\n",
        "import pandas as pd\n",
        "import qiime2 as q2\n",
        "import skbio\n",
        "from qiime2 import Artifact\n",
        "from skbio import TreeNode\n",
        "\n",
        "%load_ext autoreload\n",
        "%autoreload 2\n",
        "\n",
        "%matplotlib inline"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {},
      "source": [
        "## Fetch data"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "vscode": {
          "languageId": "shellscript"
        }
      },
      "outputs": [],
      "source": [
        "! ./../../src/fetch_mlp_data.sh"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {},
      "source": [
        "## Create Galaxy dataset"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {},
      "outputs": [],
      "source": [
        "galaxy_motus = pd.read_csv(\n",
        "    \"../../data/u3_mlp_nishijima24/GALAXY_mOTUs_v25.tsv\", sep=\"\\t\", index_col=0\n",
        ")\n",
        "galaxy_motus.head()"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {},
      "source": [
        "Number in `[ref_mOTU_v25_(\\d+)]` matches the index in file \"motus2GTDB.txt\" -> taxonomy\n",
        "\n",
        "\"-1\" column is probably undefined"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {},
      "source": [
        "No phylogeny tree can be constructed since we do not have the nucleotide sequences of these mOTUs -> no trac trainable"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {},
      "source": [
        "Original model is described here:\n",
        "https://github.com/grp-bork/CellCount_Nishijima_2024/blob/main/construct_models.R\n"
      ]
    }
  ],
  "metadata": {
    "kernelspec": {
      "display_name": "ritme_examples_prep_amplicon",
      "language": "python",
      "name": "python3"
    },
    "language_info": {
      "codemirror_mode": {
        "name": "ipython",
        "version": 3
      },
      "file_extension": ".py",
      "mimetype": "text/x-python",
      "name": "python",
      "nbconvert_exporter": "python",
      "pygments_lexer": "ipython3",
      "version": "3.9.19"
    }
  },
  "nbformat": 4,
  "nbformat_minor": 2
}
