{
  "cells": [
    {
      "cell_type": "markdown",
      "metadata": {},
      "source": [
        "# Usecase 1: Age prediction data preparation"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {},
      "source": [
        "This notebook can be run in the following conda environment:\n",
        "```shell\n",
        "mamba env create -f environment_prep_amplicon.yml\n",
        "conda activate ritme_examples_prep_amplicon\n",
        "pip install -e .\n",
        "qiime dev refresh-cache\n",
        "```\n",
        "\n",
        "Additionally, you must run the `vdb-config` tool and exit by pressing x (needed to initialize the wrapped SRA Toolkit for more information see [here](https://github.com/ncbi/sra-tools/wiki/05.-Toolkit-Configuration))\n",
        "```shell\n",
        "vdb-config -i\n",
        "```"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {},
      "source": [
        "## Setup"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 1,
      "metadata": {},
      "outputs": [],
      "source": [
        "import os\n",
        "import warnings\n",
        "\n",
        "import numpy as np\n",
        "import pandas as pd\n",
        "import qiime2 as q2\n",
        "from qiime2.plugins import demux\n",
        "\n",
        "import src.meta_proc_subr14 as proc_subr\n",
        "from src.meta_fetch import _fetch_all_supp_material, _fetch_sra_metadata, save_file\n",
        "from src.seq_fetch_n_process import (\n",
        "    cluster_wq_sequences,\n",
        "    create_phylogeny,\n",
        "    create_taxonomy,\n",
        "    fetch_sequences,\n",
        "    filter_sequences,\n",
        "    rarefy_sequences_w_fixed_seed,\n",
        ")\n",
        "from src.seq_trim import trim_sequences\n",
        "\n",
        "%load_ext autoreload\n",
        "%autoreload 2\n",
        "\n",
        "%matplotlib inline\n",
        "\n",
        "warnings.filterwarnings(\"ignore\", category=FutureWarning)"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 2,
      "metadata": {},
      "outputs": [],
      "source": [
        "######## USER INPUTS ########\n",
        "bioproject_id = \"PRJEB5482\"\n",
        "path_to_data = \"../../data/u1_subramanian14\"\n",
        "email = \"my@mail.com\"\n",
        "n_jobs = 6\n",
        "tag = \"01\"\n",
        "seed = 148\n",
        "######## END USER INPUTS #####"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 3,
      "metadata": {},
      "outputs": [],
      "source": [
        "if not os.path.exists(path_to_data):\n",
        "    os.makedirs(path_to_data)"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {},
      "source": [
        "## Fetch and process metadata"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 4,
      "metadata": {},
      "outputs": [
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "Metadata was read from file ../../data/u1_subramanian14/metadata.qza\n"
          ]
        },
        {
          "name": "stderr",
          "output_type": "stream",
          "text": [
            "/Users/adamova/miniforge3/envs/ritme_examples_prep_amplicon/lib/python3.9/site-packages/openpyxl/worksheet/_read_only.py:85: UserWarning: Unknown extension is not supported and will be removed\n",
            "  for idx, row in parser.parse():\n",
            "/Users/adamova/miniforge3/envs/ritme_examples_prep_amplicon/lib/python3.9/site-packages/openpyxl/worksheet/_read_only.py:85: UserWarning: Unknown extension is not supported and will be removed\n",
            "  for idx, row in parser.parse():\n"
          ]
        },
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "Shape before merge tab4: (996, 13)\n",
            "Shape before merge tab1: (50, 2)\n",
            "Shape after merge: (996, 14)\n"
          ]
        },
        {
          "name": "stderr",
          "output_type": "stream",
          "text": [
            "/Users/adamova/miniforge3/envs/ritme_examples_prep_amplicon/lib/python3.9/site-packages/openpyxl/worksheet/_read_only.py:85: UserWarning: Unknown extension is not supported and will be removed\n",
            "  for idx, row in parser.parse():\n"
          ]
        }
      ],
      "source": [
        "# fetch SRA metadata (takes ~3 min)\n",
        "sra_ids = pd.Series([bioproject_id], name=\"ID\")\n",
        "ids = q2.Artifact.import_data(\"NCBIAccessionIDs\", sra_ids)\n",
        "\n",
        "md_sra = _fetch_sra_metadata(path_to_data, ids, email, n_jobs)\n",
        "md_sra = proc_subr._process_sra_metadata(md_sra)\n",
        "\n",
        "# fetch supp. material\n",
        "url_supp = (\n",
        "    \"https://static-content.springer.com/esm/\"\n",
        "    \"art%3A10.1038%2Fnature13421/MediaObjects/\"\n",
        "    \"41586_2014_BFnature13421_MOESM97_ESM.xlsx\"\n",
        ")\n",
        "path2supp = _fetch_all_supp_material(path_to_data, url_supp)\n",
        "md_supp = proc_subr.process_supp_metadata(path2supp)"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 5,
      "metadata": {},
      "outputs": [
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "Saved processed metadata to: ../../data/u1_subramanian14/metadata_proc_v01.tsv\n",
            "Saved unique project IDs to: ../../data/u1_subramanian14/runids/PRJEB5482.tsv\n",
            "(2811, 38)\n"
          ]
        },
        {
          "data": {
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>experiment_id</th>\n",
              "      <th>biosample_id</th>\n",
              "      <th>bioproject_id</th>\n",
              "      <th>study_id</th>\n",
              "      <th>sample_accession</th>\n",
              "      <th>library_layout</th>\n",
              "      <th>instrument</th>\n",
              "      <th>platform</th>\n",
              "      <th>public</th>\n",
              "      <th>geo_location_name</th>\n",
              "      <th>...</th>\n",
              "      <th>health_status_at_sampling</th>\n",
              "      <th>diet_milk</th>\n",
              "      <th>diet_weaning</th>\n",
              "      <th>abx_ever</th>\n",
              "      <th>zygosity</th>\n",
              "      <th>age_months</th>\n",
              "      <th>age_months_rounded05</th>\n",
              "      <th>age_months_rounded1</th>\n",
              "      <th>study_name</th>\n",
              "      <th>study_cohort_name</th>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>id</th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>ERR500819</th>\n",
              "      <td>ERX466176</td>\n",
              "      <td>SAMEA2470685</td>\n",
              "      <td>PRJEB5482</td>\n",
              "      <td>ERP004898</td>\n",
              "      <td>ERS440148</td>\n",
              "      <td>PAIRED</td>\n",
              "      <td>Illumina MiSeq</td>\n",
              "      <td>ILLUMINA</td>\n",
              "      <td>True</td>\n",
              "      <td>Bangladesh, Dhaka, Mirpur</td>\n",
              "      <td>...</td>\n",
              "      <td>healthy</td>\n",
              "      <td>bd</td>\n",
              "      <td>False</td>\n",
              "      <td>False</td>\n",
              "      <td>no twins</td>\n",
              "      <td>0.098564</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.0</td>\n",
              "      <td>subramanian14</td>\n",
              "      <td>subramanian14</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>ERR500823</th>\n",
              "      <td>ERX466180</td>\n",
              "      <td>SAMEA2470686</td>\n",
              "      <td>PRJEB5482</td>\n",
              "      <td>ERP004898</td>\n",
              "      <td>ERS440149</td>\n",
              "      <td>PAIRED</td>\n",
              "      <td>Illumina MiSeq</td>\n",
              "      <td>ILLUMINA</td>\n",
              "      <td>True</td>\n",
              "      <td>Bangladesh, Dhaka, Mirpur</td>\n",
              "      <td>...</td>\n",
              "      <td>healthy</td>\n",
              "      <td>bd</td>\n",
              "      <td>False</td>\n",
              "      <td>False</td>\n",
              "      <td>no twins</td>\n",
              "      <td>1.511318</td>\n",
              "      <td>1.5</td>\n",
              "      <td>2.0</td>\n",
              "      <td>subramanian14</td>\n",
              "      <td>subramanian14</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>ERR500824</th>\n",
              "      <td>ERX466181</td>\n",
              "      <td>SAMEA2470687</td>\n",
              "      <td>PRJEB5482</td>\n",
              "      <td>ERP004898</td>\n",
              "      <td>ERS440150</td>\n",
              "      <td>PAIRED</td>\n",
              "      <td>Illumina MiSeq</td>\n",
              "      <td>ILLUMINA</td>\n",
              "      <td>True</td>\n",
              "      <td>Bangladesh, Dhaka, Mirpur</td>\n",
              "      <td>...</td>\n",
              "      <td>healthy</td>\n",
              "      <td>bd</td>\n",
              "      <td>False</td>\n",
              "      <td>False</td>\n",
              "      <td>no twins</td>\n",
              "      <td>2.365542</td>\n",
              "      <td>2.5</td>\n",
              "      <td>2.0</td>\n",
              "      <td>subramanian14</td>\n",
              "      <td>subramanian14</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>ERR500825</th>\n",
              "      <td>ERX466182</td>\n",
              "      <td>SAMEA2470688</td>\n",
              "      <td>PRJEB5482</td>\n",
              "      <td>ERP004898</td>\n",
              "      <td>ERS440151</td>\n",
              "      <td>PAIRED</td>\n",
              "      <td>Illumina MiSeq</td>\n",
              "      <td>ILLUMINA</td>\n",
              "      <td>True</td>\n",
              "      <td>Bangladesh, Dhaka, Mirpur</td>\n",
              "      <td>...</td>\n",
              "      <td>healthy</td>\n",
              "      <td>bd</td>\n",
              "      <td>False</td>\n",
              "      <td>False</td>\n",
              "      <td>no twins</td>\n",
              "      <td>3.384039</td>\n",
              "      <td>3.5</td>\n",
              "      <td>3.0</td>\n",
              "      <td>subramanian14</td>\n",
              "      <td>subramanian14</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>ERR500826</th>\n",
              "      <td>ERX466183</td>\n",
              "      <td>SAMEA2470689</td>\n",
              "      <td>PRJEB5482</td>\n",
              "      <td>ERP004898</td>\n",
              "      <td>ERS440152</td>\n",
              "      <td>PAIRED</td>\n",
              "      <td>Illumina MiSeq</td>\n",
              "      <td>ILLUMINA</td>\n",
              "      <td>True</td>\n",
              "      <td>Bangladesh, Dhaka, Mirpur</td>\n",
              "      <td>...</td>\n",
              "      <td>healthy</td>\n",
              "      <td>bd</td>\n",
              "      <td>False</td>\n",
              "      <td>False</td>\n",
              "      <td>no twins</td>\n",
              "      <td>3.844006</td>\n",
              "      <td>4.0</td>\n",
              "      <td>4.0</td>\n",
              "      <td>subramanian14</td>\n",
              "      <td>subramanian14</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "<p>5 rows \u00d7 38 columns</p>\n",
              "</div>"
            ],
            "text/plain": [
              "          experiment_id  biosample_id bioproject_id   study_id  \\\n",
              "id                                                               \n",
              "ERR500819     ERX466176  SAMEA2470685     PRJEB5482  ERP004898   \n",
              "ERR500823     ERX466180  SAMEA2470686     PRJEB5482  ERP004898   \n",
              "ERR500824     ERX466181  SAMEA2470687     PRJEB5482  ERP004898   \n",
              "ERR500825     ERX466182  SAMEA2470688     PRJEB5482  ERP004898   \n",
              "ERR500826     ERX466183  SAMEA2470689     PRJEB5482  ERP004898   \n",
              "\n",
              "          sample_accession library_layout      instrument  platform public  \\\n",
              "id                                                                           \n",
              "ERR500819        ERS440148         PAIRED  Illumina MiSeq  ILLUMINA   True   \n",
              "ERR500823        ERS440149         PAIRED  Illumina MiSeq  ILLUMINA   True   \n",
              "ERR500824        ERS440150         PAIRED  Illumina MiSeq  ILLUMINA   True   \n",
              "ERR500825        ERS440151         PAIRED  Illumina MiSeq  ILLUMINA   True   \n",
              "ERR500826        ERS440152         PAIRED  Illumina MiSeq  ILLUMINA   True   \n",
              "\n",
              "                   geo_location_name  ... health_status_at_sampling diet_milk  \\\n",
              "id                                    ...                                       \n",
              "ERR500819  Bangladesh, Dhaka, Mirpur  ...                   healthy        bd   \n",
              "ERR500823  Bangladesh, Dhaka, Mirpur  ...                   healthy        bd   \n",
              "ERR500824  Bangladesh, Dhaka, Mirpur  ...                   healthy        bd   \n",
              "ERR500825  Bangladesh, Dhaka, Mirpur  ...                   healthy        bd   \n",
              "ERR500826  Bangladesh, Dhaka, Mirpur  ...                   healthy        bd   \n",
              "\n",
              "          diet_weaning abx_ever  zygosity age_months age_months_rounded05  \\\n",
              "id                                                                          \n",
              "ERR500819        False    False  no twins   0.098564                  0.0   \n",
              "ERR500823        False    False  no twins   1.511318                  1.5   \n",
              "ERR500824        False    False  no twins   2.365542                  2.5   \n",
              "ERR500825        False    False  no twins   3.384039                  3.5   \n",
              "ERR500826        False    False  no twins   3.844006                  4.0   \n",
              "\n",
              "          age_months_rounded1     study_name study_cohort_name  \n",
              "id                                                              \n",
              "ERR500819                 0.0  subramanian14     subramanian14  \n",
              "ERR500823                 2.0  subramanian14     subramanian14  \n",
              "ERR500824                 2.0  subramanian14     subramanian14  \n",
              "ERR500825                 3.0  subramanian14     subramanian14  \n",
              "ERR500826                 4.0  subramanian14     subramanian14  \n",
              "\n",
              "[5 rows x 38 columns]"
            ]
          },
          "execution_count": 5,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "# merge\n",
        "md_all = md_sra.merge(md_supp, how=\"left\", on=\"sample_id\")\n",
        "md_all = proc_subr._postprocess_all_metadata(md_all)\n",
        "\n",
        "# save to file\n",
        "path_to_md = save_file(md_all, path_to_data, tag)\n",
        "\n",
        "# get number of samples\n",
        "nb_samples = md_all.shape[0]\n",
        "\n",
        "print(md_all.shape)\n",
        "md_all.head()"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {},
      "source": [
        "## Fetch and process sequences\n",
        "--> output after 5: `data/u1_subramanian14/otu_table_subr14_rar.tsv`        \n",
        "--> output after 3: `data/u1_subramanian14/otu_table_subr14_wq.qza`"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {},
      "source": [
        "Following the general approach outlined in [the original publication by Subramanian et al. 2014](https://doi.org/10.1038/nature13421), namely:\n",
        "1) fetching sequences from NCBI SRA\n",
        "2) trim sequences to at most 162 nucleotide length and overlap forward and reverse reads\n",
        "3) clustering sequences sharing >= 97% identity matched to the 13_8 99% Greengenes reference and remaining sequences were clustered de novo\n",
        "    * creating taxonomy and phylogeny for all OTUs        \n",
        "**--> feature table used by *ritme***\n",
        "4) filtering such that only OTUs present at or above a level of confident detection (0.1% relative abundance) in at least two fecal samples.\n",
        "5) rarefaction of resulting OTU table at 2'000 sequences per sample         \n",
        "    **--> feature table used by original publication**"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {},
      "source": [
        "#### 1. fetch"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 6,
      "metadata": {},
      "outputs": [
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "Analysing PRJEB5482 ...\n",
            "Imported ../../data/u1_subramanian14/runids/PRJEB5482.tsv as NCBIAccessionIDsDirFmt to ../../data/u1_subramanian14/runids/PRJEB5482.qza\n",
            "../../data/u1_subramanian14/PRJEB5482 found - not fetching again\n",
            "...finished fetching sequences of PRJEB5482!\n"
          ]
        }
      ],
      "source": [
        "fetch_sequences(n_jobs, path_to_data)"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 7,
      "metadata": {},
      "outputs": [],
      "source": [
        "# check size\n",
        "path_to_paired = os.path.join(path_to_data, \"PRJEB5482\", \"paired_reads.qza\")\n",
        "paired_reads = q2.Artifact.load(os.path.join(path_to_paired))\n",
        "\n",
        "# summarize\n",
        "path_summary = path_to_paired.replace(\".qza\", \"_summary.qzv\")\n",
        "if not os.path.isfile(path_summary):\n",
        "    (sum_paired,) = demux.actions.summarize(data=paired_reads)\n",
        "    sum_paired.save(path_summary)\n",
        "    print(f\"Saved paired summary in: {path_summary}\")"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {},
      "source": [
        "448 samples with 18858.3 median forward and reverse reads"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {},
      "source": [
        "#### 2. trim"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 8,
      "metadata": {},
      "outputs": [
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "Trimming: subramanian14...\n",
            "Trimmed sequences already exist: ../../data/u1_subramanian14/trimmed_subramanian14*.qza.\n"
          ]
        }
      ],
      "source": [
        "trim_sequences(path2md=path_to_md, path2seq=path_to_data, threads=n_jobs)"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 9,
      "metadata": {},
      "outputs": [],
      "source": [
        "# check size\n",
        "path_to_trim = os.path.join(path_to_data, \"trimmed_subramanian14.qza\")\n",
        "trimmed_reads = q2.Artifact.load(os.path.join(path_to_trim))\n",
        "\n",
        "# summarize\n",
        "path_summary = path_to_trim.replace(\".qza\", \"_summary.qzv\")\n",
        "\n",
        "if not os.path.isfile(path_summary):\n",
        "    (sum_reads,) = demux.actions.summarize(data=trimmed_reads)\n",
        "    sum_reads.save(path_summary)\n",
        "    print(f\"Saved demux summary in: {path_summary}\")"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {},
      "source": [
        "448 samples with 18349.5 median forward and reverse reads"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {},
      "source": [
        "#### 3. cluster"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 10,
      "metadata": {},
      "outputs": [
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "../../data/u1_subramanian14/gg_13_8_99_otus.qza found - not fetching again\n",
            "../../data/u1_subramanian14/subr14_seq_quality.qza found - not performing QC again\n",
            "../../data/u1_subramanian14/subr14_seqs_derep.qza found - not dereplicating again\n",
            "../../data/u1_subramanian14/otu_table_subr14_wq.qza found - not clustering again\n"
          ]
        }
      ],
      "source": [
        "cluster_wq_sequences(path_to_data=path_to_data, n_threads=n_jobs)"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 11,
      "metadata": {},
      "outputs": [
        {
          "data": {
            "text/plain": [
              "(448, 18150)"
            ]
          },
          "execution_count": 11,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "# check size\n",
        "path_to_otu = os.path.join(path_to_data, \"otu_table_subr14_wq.qza\")\n",
        "otu_table = q2.Artifact.load(os.path.join(path_to_otu))\n",
        "out_table_df = otu_table.view(pd.DataFrame)\n",
        "out_table_df.shape"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {},
      "source": [
        "##### Create taxonomy & phylogeny"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 12,
      "metadata": {},
      "outputs": [
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "../../data/u1_subramanian14/gg-13-8-99-515-806-nb-classifier.qza found - not retraining again\n",
            "../../data/u1_subramanian14/taxonomy_subr14.qza found - not creating taxonomy again\n"
          ]
        }
      ],
      "source": [
        "create_taxonomy(path_to_data=path_to_data)"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 13,
      "metadata": {},
      "outputs": [
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "../../data/u1_subramanian14/fasttree_tree_rooted_subr14.qza found - not creating phylogeny again\n"
          ]
        }
      ],
      "source": [
        "create_phylogeny(path_to_data=path_to_data, n_threads=2)"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {},
      "source": [
        "#### 4. filter"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 14,
      "metadata": {},
      "outputs": [
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "Filtering sequences ...\n",
            "Saved FeatureTable[Frequency] to: ../../data/u1_subramanian14/otu_table_subr14_filt.qza\n"
          ]
        }
      ],
      "source": [
        "filter_sequences(path_to_data=path_to_data, min_prevalence=2 / nb_samples)"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 15,
      "metadata": {},
      "outputs": [
        {
          "data": {
            "text/plain": [
              "(448, 850)"
            ]
          },
          "execution_count": 15,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "# check size\n",
        "path_to_otu = os.path.join(path_to_data, \"otu_table_subr14_filt.qza\")\n",
        "otu_table = q2.Artifact.load(os.path.join(path_to_otu))\n",
        "otu_table.view(pd.DataFrame).shape"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {},
      "source": [
        "#### 5. rarefy"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 16,
      "metadata": {},
      "outputs": [
        {
          "data": {
            "text/plain": [
              "(448, 850)"
            ]
          },
          "execution_count": 16,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "rarefied_table = rarefy_sequences_w_fixed_seed(path_to_otu=path_to_otu, seed=seed)\n",
        "\n",
        "# assert that rarefaction worked\n",
        "assert np.unique(rarefied_table.sum(axis=\"sample\"))[0] == 2000\n",
        "\n",
        "# save to file\n",
        "path_to_rar = os.path.join(path_to_data, \"otu_table_subr14_rar.tsv\")\n",
        "df_rarefied_table = rarefied_table.to_dataframe().transpose()\n",
        "df_rarefied_table.to_csv(path_to_rar, sep=\"\\t\")\n",
        "\n",
        "df_rarefied_table.shape"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {},
      "source": [
        "number of samples here: (448, 850)\n",
        "\n",
        "[note: denoising would yield even less: (333, 679)]"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {},
      "source": [
        "## Subset metadata by samples in feature tables"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 17,
      "metadata": {},
      "outputs": [
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "(2811, 38)\n"
          ]
        },
        {
          "data": {
            "text/plain": [
              "(448, 38)"
            ]
          },
          "execution_count": 17,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "# both feature tables have the same samples within\n",
        "assert df_rarefied_table.index.tolist() == out_table_df.index.tolist()\n",
        "\n",
        "print(md_all.shape)\n",
        "md_subset = md_all.loc[df_rarefied_table.index]\n",
        "\n",
        "path_to_md = os.path.join(path_to_data, \"md_subr14.tsv\")\n",
        "md_subset.to_csv(path_to_md, sep=\"\\t\")\n",
        "md_subset.shape"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {},
      "source": [
        "## Describe metadata and sequences"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 18,
      "metadata": {},
      "outputs": [
        {
          "data": {
            "text/plain": [
              "study_subcohort\n",
              "Healthy Twins & Triplets    448\n",
              "Name: count, dtype: int64"
            ]
          },
          "execution_count": 18,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "md_subset.study_subcohort.value_counts(dropna=False)"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 19,
      "metadata": {},
      "outputs": [
        {
          "data": {
            "text/plain": [
              "25"
            ]
          },
          "execution_count": 19,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "md_subset.host_id.nunique()"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {},
      "source": [
        "This study did not upload all sequences needed to reproduce their original RF training and testing, that would result in 50 unique infants. In the study's original train set, 12 healthy infants with 1,222 97%-identity OTUs were used to train the model and 25 twins & triplets and 13 singletons were then used to test it. However, only a fraction (25 out of 38 infants) from the test set were actually uploaded to SRA.\n",
        "\n",
        "With this we have 25 infants with 448 samples and 888 cleaned OTUs for modelling."
      ]
    }
  ],
  "metadata": {
    "kernelspec": {
      "display_name": "ritme_examples_prep_amplicon",
      "language": "python",
      "name": "python3"
    },
    "language_info": {
      "codemirror_mode": {
        "name": "ipython",
        "version": 3
      },
      "file_extension": ".py",
      "mimetype": "text/x-python",
      "name": "python",
      "nbconvert_exporter": "python",
      "pygments_lexer": "ipython3",
      "version": "3.9.19"
    }
  },
  "nbformat": 4,
  "nbformat_minor": 2
}
